### ISSUE TITLE
Pagination using cursors

### ISSUE THREAD

**Issue by SamiSammour** [nodkz/graphql-compose#40]
Hello there,
Thanks for this work I really find it helpful. Now I have a question.. is there anyway that I use cursors to paginate? Please note that my situation is to show a table with "count" rows (usually 10 rows per page) and with a control to go through the whole pages of the table, the data does not change much and the user can jump to any page he wants, I was thinking of a way to predict the cursor at a giving offset. is that possible?

**Comment by nodkz** [comment]
I think that Pagination and Connection algorithms are quite different.
Pagination use `limit` and `offset`. So you can easily calculate this values from required `page number`.

Connection use some `cursor` (`before` and `after` pointers) on document in DB and `first`, `last` args (for limiting data). Internally `cursor` also depends on which sorting do you apply for the connection. 

The `cursor` secrets:
- sort in connection can be proceeded only by `unique indexes` (to avoid over fetching, under fetching)
- for every document, we calculate cursor which contains fields from sorting.  E.g. for `.sort({ _id: 1, age: 1}` we get values for this fields from document and pass them to cursor `base64({_id: 34ef85...34, age: 28})`. For next document we calculate itself cursor with its data.
- when proceeding `before` and `after` arguments with the cursor, we parse values from it and pass them to where condition `.where({ _id: {$gt: 34ef85...34}, age: {$gt: 28}})`.

So if you somehow from `page number` can calculate `cursor` your task has solution, otherwise not (HEY I FOUND SOLUTION, will write in next comment)

PS. If it possible, do not use Connection for Pagination. Use `findMany` with `limit` and `skip`. Connections are for infinite lists.

**Comment by nodkz** [comment]
You may use `first` and `last` keys simultaneously.
Eg. if `PER_PAGE`=10 and you want to get 6th page, you may try such query
```
query {
  userConnection(first: 60, last: 10) {
    ...
  }
}
```
So by [connection specification](https://facebook.github.io/relay/graphql/connections.htm#sec-Pagination-algorithm) you request first 60 documents, and then get last 10 from them. Server should return records for sixth page.

**Comment by SamiSammour** [comment]
Thank you for your quick and clear response, it was really helpful.
As for the solution you proposed and in the [connection specification](https://facebook.github.io/relay/graphql/connections.htm#sec-Pagination-algorithm) they note:

> Including a value for both first and last is strongly discouraged, as it is likely to lead to confusing queries and results.

This means that I need to add my own resolver to the collection that understands the paging concept but doing it this way won't allow me to use Relay as a web framework from my website. isn't that right?

**Comment by nodkz** [comment]
> Including a value for both first and last is strongly discouraged, as it is likely to lead to confusing queries and results.

Yep, when you use `before` and `after` in conjuction, this is realy discouraged. So `first` and `last` can be used as I described above. But better for pagination do not use connections.

### About client side
> This means that I need to add my own resolver to the collection that understands the paging concept but doing it this way won't allow me to use Relay as a web framework from my website. isn't that right?

Connections in Relay is a simple extension (it gives you infinite lists functionality almost for free). But do not reject you to use what you want. So you can easily use Paging with Relay.

```js

import React, { PropTypes } from 'react';
import Relay from 'react-relay';

const PER_PAGE = 10;

class ProductList extends React.Component {
  static propTypes = {
    viewer: PropTypes.object,
    relay: PropTypes.object.isRequired,
  };

  loadPage(pageNum = 0) {
        this.props.relay.setVariables({
          offset: pageNum * PER_PAGE,
        });
  }

  render() {
    return (
      <div>
        {this.props.viewer.productList.map(({ item }) => {
          return (
            <div key={item.productId}>
              {item.name}
            </div>
          );
        })}
       <div>
         Pages: 
         <span onClick={() => { this.loadPage(1); }}>1</span>
         <span onClick={() => { this.loadPage(2); }}>2</span>
         <span onClick={() => { this.loadPage(3); }}>3</span>
       </div>
      </div>
    );
  }
}

export default Relay.createContainer(ProductList, {
  initialVariables: {
    limit: PER_PAGE,
    skip: 0,
  },
  fragments: {
    viewer: () => Relay.QL`
      fragment on Viewer {
        productList(skip: $skip, limit: $limit) {
          productId
          name
        }
      }
    `,
  },
});
```

----
### About server side

You may use existed `findMany` resolver: [live example of server response](http://graphql-compose.herokuapp.com/northwind/?query=%0A%7B%0A%20%20viewer%20%7B%0A%20%20%20%20productList(skip%3A%2050%2C%20limit%3A%2010%2C%20sort%3A%20PRODUCTID_ASC)%20%7B%0A%20%20%20%20%20%20productID%0A%20%20%20%20%20%20name%0A%20%20%20%20%7D%0A%20%20%7D%0A%7D%0A%20%20%20%20%20%20)
```js
{
  viewer {
    productList(skip: 50, limit: 10, sort: PRODUCTID_ASC) {
      productID
      name
    }
  }
}
```

Or of course, you may write your own `findByPage` resolver. 
```js
{
  viewer {
    productList(page: 6, perPage: 10, sort: PRODUCTID_ASC) {
      productID
      name
    }
  }
}
```
PR are welcome for such type of resolver 😉 for `graphql-compose-mongoose` (see [findMany](https://github.com/nodkz/graphql-compose-mongoose/blob/master/src/resolvers/findMany.js) and [findByIds](https://github.com/nodkz/graphql-compose-mongoose/blob/master/src/resolvers/findByIds.js) resolver for example).
Some hint [how to spin up local copy of `graphql-compose`](https://github.com/nodkz/graphql-compose-mongoose/issues/4#issuecomment-261262315).

**Comment by nodkz** [comment]
BTW you may be interesting this Relay example app https://github.com/nodkz/relay-northwind-app
[Live demo](https://nodkz.github.io/relay-northwind)
